### ISSUE TITLE
receipts attributes

### ISSUE THREAD

**Issue by andriux1990** [opensourcepos/opensourcepos#2877]
It would be very useful to use the attributes internally. Currently we can see them in items, receivings, sales. if you add receipts to me it would be great. we could use attributes for internal use. and that the receipt only shows what we want on the .. that the client will have

**Comment by andriux1990** [comment]
![Screenshot_2020-07-01-09-19-04-224_com android chrome](https://user-images.githubusercontent.com/29256343/86262448-1ff9eb80-bb7d-11ea-9d8f-6c137d7c1749.jpg)

add receipts here

**Comment by jekkos** [comment]
I think sales / receivings selection will have the attribute value displayed on the receipt as well? At least the data is loaded in the sale_lib when you generate it so you can use it anywhere in the template.

**Comment by odiea** [comment]
Yes. It does appear the the Attributes are showing on the Receipt
This is the Receipt
![2020-07-02 14_27_44-Window](https://user-images.githubusercontent.com/10962177/86420369-a56fbf80-bc93-11ea-9734-953a2cb77125.png)
This is the Sale
![2020-07-02 14_27_20-Window](https://user-images.githubusercontent.com/10962177/86420395-c20bf780-bc93-11ea-8246-fc9932d64ca8.png)

**Comment by objecttothis** [comment]
In general, I think we should eventually extend attributes to People (employees, customers, suppliers), but what I think you are saying is giving the option for Item Attributes to appear or not appear on an invoice or receipt.  I think this is a good idea and wouldn't likely take too much extension on what is already in place.

**Comment by jekkos** [comment]
Ok you want to control that separately from showing it in sales register then?

**Comment by andriux1990** [comment]
Hello, thank you all for paying attention to this topic

**Comment by objecttothis** [comment]
> 
> 
> Ok you want to control that separately from showing it in sales register then?

@jekkos  I can't tell if you are responding to me or @andriux1990 .  I think what @andriux1990 is saying is to add a flag in the definition of the attributes_definition table that represents whether that particular definition appears or does not appear in the Receipt.  If that is currently dictated by the Sales flag, then it's already there, but then Receipts isn't separate from Sales.  So for example on the invoice or receipt you might want to show the shipping weight attribute, but you wouldn't likely want that in the Sales Register because it would be too much information.

If you were responding to my suggestion about extending attributes to People, I think it can be the same framework, but I do think they should be differentiated so that you can't add a People Attribute to an item.  It's not likely, for example, that you would need an email address attached to an item.

**Comment by andriux1990** [comment]
> In general, I think we should eventually extend attributes to People (employees, customers, suppliers), but what I think you are saying is giving the option for Item Attributes to appear or not appear on an invoice or receipt. I think this is a good idea and wouldn't likely take too much extension on what is already in place.

this is what I am looking for, that can be selectable if you wish that the attribute could or could not be in the receipt .. it would be great that it would also be in these other places that I mention

**Comment by andriux1990** [comment]
> > Ok you want to control that separately from showing it in sales register then?
> 
> @jekkos I can't tell if you are responding to me or @andriux1990 . I think what @andriux1990 is saying is to add a flag in the definition of the attributes_definition table that represents whether that particular definition appears or does not appear in the Receipt. If that is currently dictated by the Sales flag, then it's already there, but then Receipts isn't separate from Sales. So for example on the invoice or receipt you might want to show the shipping weight attribute, but you wouldn't likely want that in the Sales Register because it would be too much information.
> 
> If you were responding to my suggestion about extending attributes to People, I think it can be the same framework, but I do think they should be differentiated so that you can't add a People Attribute to an item. It's not likely, for example, that you would need an email address attached to an item.

@objecttothis Any comments on this ... there are attributes that should be seen in the record ... but they do not need to be shown on the receipt, as they may contain internal information of the items

**Comment by objecttothis** [comment]
> 
> 
> > > > Ok you want to control that separately from showing it in sales register then?
> > > 
> > > 
> > > @jekkos I can't tell if you are responding to me or @andriux1990 . I think what @andriux1990 is saying is to add a flag in the definition of the attributes_definition table that represents whether that particular definition appears or does not appear in the Receipt. If that is currently dictated by the Sales flag, then it's already there, but then Receipts isn't separate from Sales. So for example on the invoice or receipt you might want to show the shipping weight attribute, but you wouldn't likely want that in the Sales Register because it would be too much information.
> > > If you were responding to my suggestion about extending attributes to People, I think it can be the same framework, but I do think they should be differentiated so that you can't add a People Attribute to an item. It's not likely, for example, that you would need an email address attached to an item.
> > 
> > 
> > @objecttothis Any comments on this ... there are attributes that should be seen in the record ... but they do not need to be shown on the receipt, as they may contain internal information of the items
> 
> Should be fairly straightforward to implement a flag in the attribute definition table, but you may want to make it non-binary. For example you may want to have a flag that can indicate that the attribute should appear in the receipt, invoice (for quotes), both or neither. You would need to have more than just a 1/0 combination at that point.

@jekkos on second thought, possibly a better solution, but maybe more work than it's worth, would be to convert the current definition_flag column in the attribute_definitions table to a binary flag or octets with the first octet keeping its current functionality, second being the receipt flag and third octet the flag indicating whether the attribute should appear on quotes.

We also need to think through expanding attributes to people.  If people attribute definitions will be stored on a different table then no need to make space for a flag to indicate that.  If, however it makes more sense to have all attribute definitions in one table, then we need to think through how we want to specify that an attribute is a people_attribute, and this will effect how we implement definition flags.  For example, definition flags might mean one set of things if the people_attribute flag is set and another if it's not set.

**Comment by jekkos** [comment]
Yes definitely I think it would be good to just add those options in the definition flag, to make attributes visible on receipt or invoice. Every flag can have it's own bit and then they can be OR'ed together.

I'd say for extending attributes further, maybe we should use the existing table, as that will be easy as we can reuse the existing config screen. Then we might need a new link to people table (person_id). We could just use the existing links table for this, as those joins will only be done only from a Person controller.
